sdk/core/azure-core/src/http/retry_policy.cpp

```diff
@@ -182,7 +182,5 @@ std::unique_ptr<RawResponse> Azure::Core::Http::RetryPolicy::Send(
 
     // Restore the original query parameters before next retry
     request.GetUrl().SetQueryParameters(std::move(originalQueryParameters));
-
-    ctx.ThrowIfCanceled();
```

Comment:
Why are we removing this? Consider moving this above the thread sleep line: 
```C++
if (retryAfter.count() > 0)
```

Reply:
Based on #1064.

Reply:
Sure, but this wasn't addressed: https://github.com/Azure/azure-sdk-for-cpp/issues/1064#issuecomment-741289676

Reply:
@JeffreyRichter - can you help answer this? I think we should side towards checking the cancellation more frequently, before any thread sleeps, and not just before I/O operations.

Reply:
Yes, we can check before a thread sleeps too. We could actually test if the expiration will happen while sleeping and then NOT sleep and just throw.

Reply:
> Yes, we can check before a thread sleeps too.
@JeffreyRichter the retry policies do sleep... So, should we return the cancellation check to the retry policies?
Or should we just ignore this PR ?

Reply:
Yes, should have a check just before the sleep. As I said above, if the cancellation will happen while sleeping, then we should just throw and not sleep.

Reply:
So, would this mean that we can throw before the timeout is reached right?

like
```cpp
Context context(5minutesTimeOut);

// Then from retry policy
Time sleepFor = 10minutes;
context.ThrowIfCancellIn(sleepFor);  // This might throw super early before the 5 min
Thread.sleep(sleepFor);
```

Do we need to document this?  Or just is it enough to tell in the exception that The SDK has detected a future TimeOut and decided to throw in advance?

Reply:
Good point. It just makes no sense to sleep and then cancel. So, I think we should do this and do NOT think we need to document it - let's see if any customer even notices. But, if they do, we'll just document it.

Reply:
Agreed. Checking and throwing if context is cancelled right before thread sleep is good enough.

Reply:
I don't agree with this policy.  Our design is we guarantee we won't be early but we don't guarantee cancellation happens on time.  This breaks that design.  

Also cancelling before a sleep isn't a good idea.  There are valid programs which sleep a thread and then call resume on the sleeping threads handle from another thread.  Resume decrements the suspend count and will result in the sleeping thread waking up early.

Reply:
@RickWinter we would only call the `Azure::Core::Internal::ThrowIfWillCancelAfter(ctx, retryAfter);` when we know the exact time that the thread will sleep and when we know in advance that there's no other mechanism to wake up and resume that thread.

For the case you describe, we shouldn't be checking for cancellation before sleeping.

I would also prefer to do the cancellation later than earlier, even if we sleep for a while just to cancel the request after that.  Or even if we don't even worry about cancelling before sleeping (since cancelling is just best effort).

@JeffreyRichter  @ahsonkhan  would you agree with Rick and me?  or do we want more conversation about it?

Reply:
First, I don't think whatever we decide is a big deal either way. We can err on the safer side today by just sleeping as we normally would and then throw if canceled. This was we never cancel early but we block a thread for no good reason (which is less efficient but is not wrong).

If any customers have issue with this, then we could shrink the sleep time to the sooner of it & cancelation. 
If any customer have issue with this, then we could avoid the sleep entirely and cancel early.
My point being that we can wait for customer feedback on this and just do the simple implementation for now.

Reply:
I agree with what @RickWinter suggested and do the simpler thing of checking once before sleeping if context has been cancelled, and that's it. Trying to calculate the subsequent sleep time and pre-emptively cancelling is error prone and unexpected, even though there is a non-optimal behavior. Ideally, we use thread sleep sparingly and for really small durations.

As @JeffreyRichter stated:
> and just do the simple implementation for now.

To clarify about these:
> I would also prefer to do the cancellation later than earlier, even if we sleep for a while just to cancel the request after that.

>  We can err on the safer side today by just sleeping as we normally would and then throw if canceled.

I agree, but check for cancellation and throw **before** sleeping. We can also check after, if desired, but let's **not** only check after.

Reply:
Reverted the feature to check if will be cancelled. :)